Add a Kogito (Drools + jBPM) extension (#2775)
quarkusio/quarkus: docs/src/main/asciidoc/kogito-guide.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////

include::./attributes.adoc[]
= {project-name} - Using Kogito to add business automation capabilities to an application

This guide demonstrates how your {project-name} application can use Kogito to add business automation
to power it up with business processes and rules.

> Kogito is a next generation business automation toolkit that originates from well know Open Source projects

Reviewer: ```suggestion
Kogito is a next generation business automation toolkit that originates from well known Open Source projects
```

Drools (for business rules) and jBPM (for business processes). Kogito aims at providing another approach

> to business automation where the main message is to expose you business knowledge (processes, rules and decisions)

Reviewer: ```suggestion
to business automation where the main message is to expose your business knowledge (processes, rules and decisions)
```

> in domain specific way.

Reviewer: ```suggestion
in a domain specific way.
```
?

> You can find more information about Kogito in https://kogito.kie.org/[Kogito website].

Reviewer: ```suggestion
You can find more information about Kogito in the https://kogito.kie.org/[Kogito website].
```

== Prerequisites

To complete this guide, you need:

* less than 15 minutes

> * an IDE (preferred Eclipse with BPMN modeller plugin)

Reviewer: ```suggestion
* an IDE (Eclipse is preferred with the BPMN modeller plugin)
```

* JDK 1.8+ installed with `JAVA_HOME` configured appropriately
* Apache Maven 3.5.3+
* Docker

=== Install Eclipse with modelling plugins

> To be able to make use of visual modelling of your processes download Eclipse IDE and

Reviewer: ```suggestion
To be able to make use of visual modelling of your processes, download Eclipse IDE and
```

install from Market place

> * Eclipse BPMN2 Modeler plugin (with jBPM Runtime Extension)

Reviewer: Why use a bullet list here if there is a single plugin? It looks a bit weird.

== Architecture

In this example, we build a very simple microservice which offers one REST endpoint:

* `/persons`

This endpoint will be automatically generated based on business process, that in turn will

> make use of business rules to make certain decisions based on the data being processed by the process.

Reviewer: Maybe just `being processed`? `being processed by the process` is a bit redundant.

=== Business process

> Business process will be responsible for encapsulating business logic of our microservice.

Reviewer: ```suggestion
The business process will be responsible for encapsulating business logic of our microservice.
```

> It should provide complete set of steps to achieve particular business goal.

Reviewer: ```suggestion
It should provide complete set of steps to achieve particular a business goal.
```
(or plural, as you prefer)

At the same time this is the entry point to the service that can be consumed by clients.

=== Business rule

> Business rule allows to externalise decision logic into reusable pieces that can be easily

Reviewer: ```suggestion
A business rule allows to externalise decision logic into reusable pieces that can be easily
```

that comes with all needed dependencies and configuration to equip your application
with business automation.

== Writing the application

Let's start by implementing the simple data object `Person`. As you can see from the source code below it is just a POJO:

[source,java]
----
package org.acme.kogito.model;

public class Person {

	private String name;
	private int age;
	private boolean adult;

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public int getAge() {
		return age;
	}

	public void setAge(int age) {
		this.age = age;
	}

	public boolean isAdult() {
		return adult;
	}

	public void setAdult(boolean adult) {
		this.adult = adult;
	}

	@Override
	public String toString() {
		return "Person [name=" + name + ", age=" + age + ", adult=" + adult + "]";
	}

}

----

Next, we create a rule file `person-rules.drl` inside the `src/main/resources/org/acme/kogito` folder of
the generated project.

[source,plain]
----
package org.acme.kogito

import org.acme.kogito.model.Person;

rule "Is adult" ruleflow-group "person"

when
    $person: Person(age > 18)
then
    modify($person) {
    	setAdult(true)
    };
end
----

> This is really a simple rule that marks person who is older that 18 years as an adult.

Reviewer: ```suggestion
This is really a simple rule that marks a person who is older that 18 years as an adult.
```

Finally we create a business process that will make use of this rule and some other
activities to approve a given person. Using eclipse wizard create `persons.bpmn` inside
 `src/main/resources/org/acme/kogito` folder of the generated project.

> This process should consists of

Reviewer: ```suggestion
This process should consist of
```

* start event
* business rule task
* exclusive gateway
* user task
* end events

And should look like

image:kogito-guide-screenshot.png[alt=Architecture]

> To get started quickly copy the process definition from

Reviewer: ```suggestion
To get started quickly copy the process definition from the
```

{quickstarts-tree-url}/using-kogito/src/main/resources/org/acme/kogito/persons.bpmn2[quickstart]

To model this process yourself, just follow these steps (start event should be automatically added)

> * define process variable with name `person` of type `org.acme.kogito.model.Person`

Reviewer: ```suggestion
* define a process variable with name `person` of type `org.acme.kogito.model.Person`
```

* drag the Tasks -> Business Rule Task from the palette and drop it next to start event, link it with start event
** double click on the business rule task
*** on tab I/O Parameters, set data input and output (map `person` process variable to input data with name `person` and same for data output)

> *** on tab Business Rule Task, set rule flow group to the value defined in drl file (`person`)

Reviewer: ```suggestion
*** on tab Business Rule Task, set rule flow group to the value defined in the drl file (`person`)
```

* drag the Tasks -> User Task from the palette and drop it next to the gateway, link it with gateway
** double click on the user task
*** on tak User Task, set task name to `ChildrenHandling`
*** on tab I/O Parameters, set data input (map `person` process variable to input data with name `person`)
* drag the End Events -> End from the palette and drop it next to the user task, link it with the user task
* drag the End Events -> End from the palette and drop it next to the gateway, link it with the user task
* double click on the gateway
** on tab Gateway, set the diverging direction for the gateway
** on tab Gateway, set conditions on sequence flow list
*** -> going to end event `return person.isAdult() == true;` with language `Java`
*** -> going to user task `return person.isAdult() == false;` with language `Java`
* save the file

== Running and Using the Application

=== Running in Developer Mode

To run the microservice in dev mode, use `./mvnw clean compile quarkus:dev`.

=== Running in JVM Mode

When you're done playing with "dev-mode" you can run it as a standard Java application.

First compile it:

[source,bash]
----

> mvn package

Reviewer: Please use `./mvnw` here too.

----

Then run it:

[source,bash]
----
java -jar ./target/using-kogito-runner.jar
----

=== Running in Native Mode

This same demo can be compiled into native code: no modifications required.

This implies that you no longer need to install a JVM on your
production environment, as the runtime technology is included in
the produced binary, and optimized to run with minimal resource overhead.

Compilation will take a bit longer, so this step is disabled by default;
let's build again by enabling the `native` profile:

[source,bash]
----

> mvn package -Dnative

Reviewer: Same here about `./mvnw`

----

After getting a cup of coffee, you'll be able to run this binary directly:

[source,bash]
----
./target/using-kogito-runner
----

== Testing the Application

> To test your application it, just send request to the service with giving the person as JSON

Reviewer: ```suggestion
To test your applicatio, just send a request to the service with giving the person as JSON
```

payload.

[source,bash]
----
curl -X POST http://localhost:8080/persons \
    -H 'content-type: application/json' \
    -H 'accept: application/json' \
    -d '{"person": {"name":"John Quark", "age": 20}}'
----

> In response person should be approved as adult and that should also be visible in response payload.

Reviewer: ```suggestion
In the response, the person should be approved as an adult and that should also be visible in the response payload.
```

[source,JSON]
----
{"id":1,"person":{"adult":true,"age":20,"name":"John Quark"}}
----

You can also verify that there are no more active instances

[source,bash]
----
curl -X GET http://localhost:8080/persons \
    -H 'content-type: application/json' \
    -H 'accept: application/json'
----

> To verify the non adult case, send another request with age less than 18

Reviewer: ```suggestion
To verify the non adult case, send another request with the age set to less than 18
```

[source,bash]
----
curl -X POST http://localhost:8080/persons \
    -H 'content-type: application/json' \
    -H 'accept: application/json' \
    -d '{"person": {"name":"Jenny Quark", "age": 15}}'
----

this time there should be one active instance

[source,bash]
----
curl -X GET http://localhost:8080/persons/2/tasks \
    -H 'content-type: application/json' \
    -H 'accept: application/json'
----

> You can get details of the task by calling another endpoint

Reviewer: ```suggestion
You can get the details of the task by calling another endpoint
```

[source,bash]
----
curl -X GET http://localhost:8080/persons/2/ChildrenHandling/1 \
    -H 'content-type: application/json' \
    -H 'accept: application/json'
----

> You can complete this person evaluation process instance by calling same endpoint but with POST

Reviewer: ```suggestion
You can complete this person evaluation process instance by calling the same endpoint but with a POST request
```

decision trees, rules, etc. For this example we focus on the rule format backed by DRL
(Drools Rule Language).

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can go right to the complete example.

Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].

The solution is located in the `using-kogito` {quickstarts-tree-url}/using-kogito[directory].

== Creating the Maven Project

First, we need a new project. Create a new project with the following command:

[source, subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=using-kogito \
    -Dextensions="kogito"
----

This command generates a Maven project, importing the `kogito` extension
that comes with all needed dependencies and configuration to equip your application
with business automation.

== Writing the application

Let's start by implementing the simple data object `Person`. As you can see from the source code below it is just a POJO:

[source,java]
----
package org.acme.kogito.model;

> import java.io.Serializable;

Reviewer 2: that's to support process variables to be persisted for long running process instances

> import java.io.Serializable;

Reviewer 3: ok, thanks!

> import java.io.Serializable;

Reviewer 3: Be aware that it may not work in native (@dmlloyd can tell you more about it)

> import java.io.Serializable;

Reviewer 2: good to know about potential issues with serialisation, I can remove it fro the time being

public class Person implements Serializable {

	private static final long serialVersionUID = -571683427125356701L;

	private String name;
	private int age;
	private boolean adult;

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public int getAge() {
		return age;
	}

	public void setAge(int age) {
		this.age = age;
	}

	public boolean isAdult() {
		return adult;
	}

	public void setAdult(boolean adult) {
		this.adult = adult;
	}

	@Override
	public String toString() {
		return "Person [name=" + name + ", age=" + age + ", adult=" + adult + "]";
	}

}

----

Next, we create a rule file `person-rules.drl` inside the src/main/resources/org/acme/kogito folder of
the generated project.

[source,plain]
----
package org.acme.kogito

import org.acme.kogito.model.Person;

rule "Is adult" ruleflow-group "person"

when
    $person: Person(age > 18)
then
    modify($person) {
    	setAdult(true)
    };
end
----

This is really a simple rule that marks person who is older that 18 years as an adult.

Finally we create a business process that will make use of this rule and some other

> activities to approve a given person. Using eclipse wizard create `persons.bpmn` inside

Reviewer 2: it's a new item (class, files, process) wizard, will make it more clear

> activities to approve a given person. Using eclipse wizard create `persons.bpmn` inside

Reviewer: Yes, please. I agree with @cescoffier that it's not clear enough.
